File: frontend/src/api/anomalies.ts
/**
 * Anomaly Detection API client
 * Cross-domain anomaly scanning for investigative analysts
 */
import { apiClient } from './client'

// ── Enums ──────────────────────────────────────────────────────

export type AnomalySeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW'
export type AnomalyType = 'same_day_cluster' | 'rapid_director_change' | 'non_filer_cluster' | 'pan_anomaly'

// ── Base ───────────────────────────────────────────────────────

export interface AnomalyBase {
  type: AnomalyType
  severity: AnomalySeverity
  title: string
  description: string
  entities: string[]
}

// ── Same-Day Clusters ──────────────────────────────────────────

export interface SameDayCompany {
  id: string
  name_english: string
  registration_number: number
  pan: string | null
  company_type_category: string | null
}

export interface SameDayCluster extends AnomalyBase {
  type: 'same_day_cluster'
  registration_date: string
  address: string
  company_count: number
  companies: SameDayCompany[]
}

// ── Rapid Director Changes ─────────────────────────────────────

export interface RapidDirectorChange extends AnomalyBase {
  type: 'rapid_director_change'
  company_id: string
  company_name: string
  director_name: string
  director_role: string | null
  appointed_date: string | null
  resigned_date: string | null
  duration_days: number
}

// ── Non-Filer Clusters ─────────────────────────────────────────

export interface NonFilerClusterCompany {
  id: string
  name_english: string
  pan: string | null
  is_non_filer: boolean
}

export interface NonFilerCluster extends AnomalyBase {
  type: 'non_filer_cluster'
  address: string
  total_companies: number
  non_filer_count: number
  non_filer_pct: number
  companies: NonFilerClusterCompany[]
}

// ── PAN Anomalies ──────────────────────────────────────────────

export interface PANAnomalyCompany {
  id: string
  name_english: string
  registration_number: number
  company_address: string | null
}

export interface PANAnomaly extends AnomalyBase {
  type: 'pan_anomaly'
  pan: string
  company_count: number
  companies: PANAnomalyCompany[]
}

// ── Summary & Full Scan ────────────────────────────────────────

export interface AnomalySummary {
  same_day_clusters: number
  rapid_director_changes: number
  non_filer_clusters: number
  pan_anomalies: number
  total: number
}

export interface AnomalyScanResult {
  summary: AnomalySummary
  same_day_clusters: SameDayCluster[]
  rapid_director_changes: RapidDirectorChange[]
  non_filer_clusters: NonFilerCluster[]
  pan_anomalies: PANAnomaly[]
  scanned_at: string | null
}

// ── API Functions ──────────────────────────────────────────────

export async function getAnomalySummary(): Promise<AnomalySummary> {
  const { data } = await apiClient.get('/anomalies/summary')
  return data
}

export async function getSameDayClusters(minCount: number = 3, limit: number = 100): Promise<SameDayCluster[]> {
  const { data } = await apiClient.get('/anomalies/same-day-clusters', {
    params: { min_count: minCount, limit },
  })
  return data
}

export async function getRapidDirectorChanges(maxDays: number = 90, limit: number = 200): Promise<RapidDirectorChange[]> {
  const { data } = await apiClient.get('/anomalies/rapid-director-changes', {
    params: { max_days: maxDays, limit },
  })
  return data
}

export async function getNonFilerClusters(minPct: number = 60, limit: number = 100): Promise<NonFilerCluster[]> {
  const { data } = await apiClient.get('/anomalies/non-filer-clusters', {
    params: { min_pct: minPct, limit },
  })
  return data
}

export async function getPANAnomalies(minCompanies: number = 5, limit: number = 100): Promise<PANAnomaly[]> {
  const { data } = await apiClient.get('/anomalies/pan-anomalies', {
    params: { min_companies: minCompanies, limit },
  })
  return data
}

export async function runFullScan(): Promise<AnomalyScanResult> {
  const { data } = await apiClient.get('/anomalies/full-scan')
  return data
}

File: backend-v5/app/api/v1/anomalies.py
"""Anomaly Detection API endpoints -- cross-domain anomaly scanning for investigative analysts."""
from typing import List

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import require_analyst
from app.core.database import get_db
from app.models.user import User
from app.services.anomaly_detection_service import AnomalyDetectionService
from app.schemas.anomaly import (
    AnomalySummary,
    SameDayCluster,
    RapidDirectorChange,
    NonFilerCluster,
    PANAnomaly,
    AnomalyScanResult,
)

router = APIRouter(prefix="/anomalies", tags=["Anomaly Detection"])


# ============================================================
# Summary (fast counts for dashboard badges)
# ============================================================

@router.get("/summary", response_model=AnomalySummary)
async def get_anomaly_summary(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_analyst),
):
    """
    Aggregate counts of all anomaly types for dashboard summary badges.

    Returns counts for: same-day registration clusters, rapid director changes,
    non-filer address clusters, and PAN anomalies.
    Analyst+ access required.
    """
    service = AnomalyDetectionService(db)
    return await service.get_anomaly_summary()


# ============================================================
# Same-Day Registration Clusters
# ============================================================

@router.get("/same-day-clusters", response_model=List[SameDayCluster])
async def get_same_day_clusters(
    min_count: int = Query(default=3, ge=2, le=50, description="Minimum companies at same address on same day"),
    limit: int = Query(default=100, ge=1, le=500, description="Maximum clusters to return"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_analyst),
):
    """
    Find dates where min_count+ companies registered at the same address on
    the same day. A strong signal for coordinated shell company creation.

    Analyst+ access required.
    """
    service = AnomalyDetectionService(db)
    return await service.detect_same_day_registration_clusters(
        min_count=min_count,
        limit=limit,
    )


# ============================================================
# Rapid Director Changes
# ============================================================

@router.get("/rapid-director-changes", response_model=List[RapidDirectorChange])
async def get_rapid_director_changes(
    max_days: int = Query(default=90, ge=1, le=365, description="Maximum days between appointment and resignation"),
    limit: int = Query(default=200, ge=1, le=1000, description="Maximum results to return"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_analyst),
):
    """
    Companies where directors were appointed AND resigned within max_days days.
    Rapid turnover indicates nominee directors or fraudulent arrangements.

    Analyst+ access required.
    """
    service = AnomalyDetectionService(db)
    return await service.detect_rapid_director_changes(
        max_days=max_days,
        limit=limit,
    )


# ============================================================
# Non-Filer Clusters
# ============================================================

@router.get("/non-filer-clusters", response_model=List[NonFilerCluster])
async def get_non_filer_clusters(
    min_pct: float = Query(default=60.0, ge=10.0, le=100.0, description="Minimum non-filer percentage at address"),
    limit: int = Query(default=100, ge=1, le=500, description="Maximum clusters to return"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_analyst),
):
    """
    Addresses where the majority (>min_pct%) of companies are IRD non-filers.
    High non-filer concentration suggests potential tax avoidance clusters.

    Analyst+ access required.
    """
    service = AnomalyDetectionService(db)
    return await service.detect_non_filer_clusters(
        min_pct=min_pct,
        limit=limit,
    )


# ============================================================
# PAN Anomalies
# ============================================================

@router.get("/pan-anomalies", response_model=List[PANAnomaly])
async def get_pan_anomalies(
    min_companies: int = Query(default=5, ge=2, le=100, description="Minimum companies per PAN to flag"),
    limit: int = Query(default=100, ge=1, le=500, description="Maximum anomalies to return"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_analyst),
):
    """
    PANs linked to min_companies+ companies. A single PAN controlling many
    companies is unusual and warrants investigation.

    Analyst+ access required.
    """
    service = AnomalyDetectionService(db)
    return await service.detect_pan_anomalies(
        min_companies=min_companies,
        limit=limit,
    )


# ============================================================
# Full Scan (all detectors)
# ============================================================

@router.get("/full-scan", response_model=AnomalyScanResult)
async def run_full_scan(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_analyst),
):
    """
    Run all anomaly detectors and return combined results.
    This is a heavier query -- use the individual endpoints for targeted analysis.

    Analyst+ access required.
    """
    service = AnomalyDetectionService(db)
    return await service.run_full_scan()
